mkdocs_git_authors_plugin/plugin.py

```diff
@@ -1,15 +1,81 @@
 import re
 from mkdocs.config import config_options
 from mkdocs.plugins import BasePlugin
-from .util import Util
+from .repo import Repo
 
 class GitAuthorsPlugin(BasePlugin):
     config_scheme = (
         ('show_contribution', config_options.Type(bool, default=False)),
+        ('show_lines', config_options.Type(bool, default=False)),
+        ('count_empty_lines', config_options.Type(bool, default=True)),
+        ('label_lines', config_options.Type(str, default='lines')),
+        ('sort_by', config_options.Choice(
+            ['name', 'contribution'], default='name')
+        ),
+        ('sort_reverse', config_options.Type(bool, default=False)),
+        ('uncommitted_name', config_options.Type(str, default='Uncommitted')),
+        ('uncommitted_email', config_options.Type(str, default='#'))
     )
 
     def __init__(self):
-        self.util = Util()
+        self._repo = Repo()
+
+    def on_config(self, config, **kwargs):
+        """
+        Store the plugin configuration in the Repo object.
+
+        This is only the dictionary with the plugin configuration,
+        not the global config which is passed to the various event handlers.
+        """
+        self.repo().set_config(self.config)
+
+    def on_files(self, files, **kwargs):
+        """
+        Preprocess all markdown pages in the project
+
+        This populates all the lines and total_lines properties
+        of the pages and the repository, so the total
+        contribution of an author to the repository can be
+        retrieved on *any* Markdown page.
+        """
+        for file in files:
+            path = file.abs_src_path
+            if path.endswith('.md'):
+                _ = self.repo().page(path)
+
+    def on_page_content(self, html, page, config, files, **kwargs):
+        """
+        Replace jinja tag {{ git_authors_list }} in HTML.
```

Comment:
Let's discuss, because it's a big design decision for the plugin. My considerations:

Because I'm thinking about future integration with `git-revision-date`: 

1) I would like to keep it simple by having only one ``{{ git_authors_summary }}` tag. 
2) I would also like to keep the number of options down

Because I would like to promote integration into existing themes (most prominently `mkdocs-material`):

3) Have all the potentially relevant information exposed in `{{ git_authors }}` 
4) Promote users that want to do more than the simple summary to customize their used themes. (I updated the README to explain how to do this for `mkdocs-material` in 0668d3e51176a414e3a9410df4a79a26c31ab935)

So yes, being able to make an HTML list of authors is definitely in scope of the plugin, but we have not aligned yet on *how* do this properly.

Reply:
OK, I see. Please let this go through the PR and let us discuss this later, but have it clearly resolved in a way you're happy with before pushing a new release to PyPi. Right now I'm the only user of that functionality, and that will remain the case until it is available from PyPi. And I'm of course ready to adapt my few projects to whatever solution this will lead. 

I think the current state of it doesn't have serious side effects, it is "only" an additional function you don't know yet whether to keep. So I don't think there's much of a risk merging it temporarily.

Reply:
I've created #22 in order to mark this as resolved. Of course you can reopen it if you think it's necessary to add more discussion *here*.

Reply:
This is tough. I understand you have already implemented this PR in several projects.  

> The first rule of open source, according to `@shykes`: “No is temporary, yes is forever.”

I do not want to introduce extra markdown tags until we have a clear direction on how to deal with customization (I created https://github.com/timvink/mkdocs-git-authors-plugin/issues/21 for this), not even on unreleased master branch. 

I will merge this PR, and comment out `plugin.Plugin.on_page_content()`. You can fork, uncomment, and use that fork in your packages while we decide on #21.